src/rules/switchDefaultRule.ts

```diff
@@ -15,41 +15,216 @@
+
+                if (uncovered === "non-exhaustive") {
+                    ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+                } else if (uncovered.length > 0) {
+                    ctx.addFailureAtNode(node, Rule.failureStringForMissingCases(uncovered));
+                }
+            }
         }
         return ts.forEachChild(node, cb);
     });
 }
+
+// returns a (possibly empty) array of missing values, or the string 'non-exhaustive' if the
+// type of the switched expression can't possibly be matched exhaustively
+function uncoveredValues(
+    checker: ts.TypeChecker,
+    switchStatement: ts.SwitchStatement,
+): "non-exhaustive" | SingletonValue[] {
+    const switchedExprType = checker.getTypeAtLocation(switchStatement.expression);
+
+    const possibleValues: SingletonValue[] = [];
+    for (const type of unionTypeParts(switchedExprType)) {
+        // Note: `boolean` is represented as `true | false` under the hood, so we don't have to
+        // do anything extra to support it
+        const value = singletonValueOfType(type);
+
+        if (value === undefined) {
+            return "non-exhaustive";
+        }
+
+        possibleValues.push(value.value);
+    }
+
+    const caseClauses = switchStatement.caseBlock.clauses as ts.NodeArray<ts.CaseClause>;
+
+    const caseValues: SingletonValue[] = [];
+    for (const clause of caseClauses) {
+        const type = checker.getTypeAtLocation(clause.expression);
+        const value = singletonValueOfType(type);
+
+        if (value === undefined) {
+            return "non-exhaustive";
+        }
+
+        caseValues.push(value.value);
+    }
+
+    // PseudoBigInt is an object, and indexOf will use reference equality. But this should be OK
+    // because AFAICT TypeScript interns all literal types in a Map, and if the type checker returns
+    // the same type twice, it will be the same object (and therefore that object's `value` field
+    // will be the same, too).
+    return possibleValues.filter(value => caseValues.indexOf(value) === -1);
+}
+
+// ts.PseudoBigInt was added in Typescript 3.2, and is a possible `LiteralType` value.
+// Since it is not available in earlier versions, we copy its definition and use it here.
+interface PseudoBigInt {
+    negative: boolean;
+    base10Value: string;
+}
+
+type SingletonValue = string | number | PseudoBigInt | true | false | null | undefined;
+
+// if `type` is a singleton type (a literal, `true`, `false`, `null`, or `undefined`),
+// returns an object whose `value` field is the single value that occupies that type.
+// Otherwise, returns `undefined`.
+// The object wrapper is to distinguish between `undefined` and `{value: undefined}`.
+function singletonValueOfType(type: ts.Type): { value: SingletonValue } | undefined {
+    if (isTypeFlagSet(type, ts.TypeFlags.Null)) {
+        return { value: null };
+    } else if (isTypeFlagSet(type, ts.TypeFlags.Undefined)) {
```

Comment:
Nit: you can remove the `else` if the preceding `if` block `return`s. Makes it a bit easier to read 🙂